api/src/main/java/jakarta/data/repository/KeysetPageable.java

```diff
@@ -0,0 +1,265 @@
+ * &#64;OrderBy("zipcode")
+ * &#64;OrderBy("birthYear")
+ * &#64;OrderBy("id")
+ * KeysetAwareSlice&lt;Customer&gt; getTopBuyers(int minOrders, float minSpent, Pageable pagination);
+ * </pre>
+ *
+ * <h2>Page Numbers and Totals</h2>
+ *
+ * <p>Page numbers, total numbers of elements across all pages, and total
+ * count of pages are not accurate when keyset pagination is used and should
+ * not be relied upon.</p>
+ */
+public class KeysetPageable extends Pageable {
+    /**
+     * Direction of keyset pagination.
+     */
+    public static enum Mode {
+        /**
+         * Indicates forward keyset pagination, which follows the
+         * direction of the {@link OrderBy} annotations, <code>Pageable</code>
+         * {@link Pageable#of(long, long, Sort...) Sort} parameters,
+         * or <code>OrderBy</code> name pattern of the repository method.
+         */
+        NEXT,
+
+        /**
+         * Indicates reverse keyset pagination. which follows the opposite
+         * direction of the {@link OrderBy} annotations, <code>Pageable</code>
+         * {@link Pageable#of(long, long, Sort...) Sort} parameters,
+         * or <code>OrderBy</code> name pattern of the repository method.
+         */
+        PREVIOUS
+    }
+
+    /**
+     * Represents keyset values, which can be a starting point for
+     * requesting a next or previous page.
+     */
+    public static class Cursor {
+        private final Object[] keyset;
+
+        private Cursor(Object[] keyset) {
+            this.keyset = keyset;
+        }
+
+        /**
+         * Returns the keyset value at the specified position.
+         *
+         * @param  index position (0 is first) of the keyset value to obtain.
+         * @return the keyset value at the specified position.
+         * @throws IndexOutOfBoundsException if the index is negative
+         *         or greater than or equal to the {@link #size}.
+         */
+        public Object getKeysetElement(int index) {
+            return keyset[index];
+        }
+
+        /**
+         * Returns the number of values in the keyset.
+         *
+         * @return the number of values in the keyset.
+         */
+        public int size() {
+            return keyset.length;
+        }
+
+        /**
+         * String representation of the keyset cursor, including the number of
+         * key values in the cursor but not the values themselves.
+         *
+         * @return String representation of the keyset cursor.
+         */
+        @Override
+        public String toString() {
+            return new StringBuilder("Cursor@").append(Integer.toHexString(hashCode())) //
+                            .append(" with ").append(keyset.length).append(" keys") //
+                            .toString();
+        }
+    }
+
+    private final Cursor cursor;
+    private final Mode mode;
+
+    KeysetPageable(Pageable copyFrom, Mode mode, Object... keyset) {
```

Comment:
IMO we should accept a `Cursor` instead of `Object...` here for the keyset here and also add methods `Pageable#afterKeyset(Cursor)`/`Pageable#beforeKeyset(Cursor)`, as injecting a custom cursor might be desirable.